Translate reconciliation (#41)
reactjs/pt-BR.reactjs.org: content/docs/reconciliation.md

---
id: reconciliation
title: Reconciliação (Reconciliation)
permalink: docs/reconciliation.html
---

O React provê uma API declarativa, assim, você não precisa se preocupar em saber exatamente o que mudou em cada atualização. Isso torna mais fácil a criação de aplicações, apesar de não ser óbvia a forma como isso é implementado no React. Este artigo explica as escolhas que fizemos no algoritmo de _diffing_ para que as atualizações nos componentes sejam previsíveis e rápidas o suficiente para aplicações de alta performance.

## Motivação {#motivation}

> Quando utilizamos React, podemos imaginar a função `render()` como uma função responsável por gerar à árvore de Elementos React. Na próxima atualização de `state` ou `props`, a função `render()` retornará uma árvore de Elementos React diferente. Dessa forma, o React precisará descobrir como ele pode atualizar a UI para sincronizar com a árvore mais recente.

Reviewer: Remover a crase de **`gerar à arvore`**

Existem algumas soluções genéricas para o problema deste algoritmo de gerar o menor número de operações necessário para transformar uma árvore em outra. Contudo, o [algoritmo de última geração](http://grfia.dlsi.ua.es/ml/algorithms/references/editsurvey_bille.pdf) possui uma complexidade da ordem de O(n<sup>3</sup>), onde n é o numero de elementos na àrvore.

Se usássemos isso no React, exibir 1000 elementos iria requerer aproximadamente um bilhão de comparações. Isso é, de longe, muito custoso. Em vez disso, o React implementa um algoritmo heurístico  da ordem de O(n) baseado em duas suposições:

1. Dois elementos de tipos diferentes irão produzir árvores diferentes.
2. O desenvolvedor pode indicar quais elementos filhos estão estáveis entre diferentes renderizações através da propriedade `key`

Na prática, essas suposições são válidas para quase todos os casos práticos.

> ## O Algoritmo de Diferenciação {#the-diffing-algorithm}

Reviewer 2: ```suggestion
## O Algoritmo de Diferenciação (_Diffing_) {#the-diffing-algorithm}
```

Quando diferenciando duas árvores, o React primeiro compara os dois elementos raíz. O comportamento é diferente dependendo do tipo dos elementos raíz.

### Elementos de Tipos Diferentes {#elements-of-different-types}

Sempre que os elementos raíz tiverem tipos diferentes, o React irá destruir a árvore antiga e construir uma árvore nova do zero. Indo de  `<a>` para `<img>`, ou de `<Article>` para `<Comment>`, ou de `<Button>` para `<div>` - qualquer uma dessas mudanças resultará em uma reconstrução total.

> Quando desruímos uma árvore, os nós antigos do DOM são destruídos. instâncias de componentes recebem `componentWillUnmount()`. Quando construímos uma nova árvore, novos nós do DOM são inseridos no DOM. instâncias de componentes recebem `componentWillMount()` and depois `componenteDidMount()`. Qualquer estado associado com a árvore antiga é perdido.

Reviewer 2: ```suggestion
Quando destruímos uma árvore, os nós antigos do DOM são destruídos. Instâncias de componentes recebem `componentWillUnmount()`. Quando construímos uma nova árvore, novos nós do DOM são inseridos no DOM. Instâncias de componentes recebem `componentWillMount()` e depois `componentDidMount()`. Qualquer estado associado com a árvore antiga é perdido.
```

---
id: reconciliation

> title: Reconciliação

Reviewer 3: let's keep `Reconciliation` here as it's a common word in React

```suggestion
title: Reconciliação (Reconciliation)
```

permalink: docs/reconciliation.html
---

O React provê uma API declarativa, assim, você não precisa se preocupar em saber exatamente o que mudou aem cada atualização. Isso torna mais fácil a criação de aplicações, apesar de não ser obvia a forma como isso é implementado no React. Este artigo explica as escolhas que fizemos no algoritimo de "diffing" para que as atualizaçãos nos componentes sejam previsíveis e rapidas o suficiente para aplicações de alta performance.

## Motivação {#motivation}

Quando utilizamos React, podemos imaginar a função `render()` como uma função responsável por gerar à árvore de Elementos React.

> Na proxima atualização de `state` ou `props`, a função `render()` retornará uma árvore de Elementos React diferente. Dessa forma, o React precisará descobrir como ele pode atualizar a UI para sincronizar com a arvore mais recente.

Reviewer 3: s/proxima/próxima

> Na proxima atualização de `state` ou `props`, a função `render()` retornará uma árvore de Elementos React diferente. Dessa forma, o React precisará descobrir como ele pode atualizar a UI para sincronizar com a arvore mais recente.

Reviewer 3: also please keep this text in the same line as the above text to avoid breaking stuff if an update comes from the upstream repo

> Existem algumas soluções genéricas para o problema deste algoritmo de gerar menor número de operações para transformar uma árvore em outra. Contudo, o [algoritmo de última geração](http://grfia.dlsi.ua.es/ml/algorithms/references/editsurvey_bille.pdf) possui uma complexidade da ordem de O(n<sup>3</sup>), onde n é o numero de elementos na àrvore.

Reviewer 3: ```suggestion
Existem algumas soluções genéricas para o problema deste algoritmo de gerar o menor número de operações necessário para transformar uma árvore em outra. Contudo, o [algoritmo de última geração](http://grfia.dlsi.ua.es/ml/algorithms/references/editsurvey_bille.pdf) possui uma complexidade da ordem de O(n<sup>3</sup>), onde n é o numero de elementos na àrvore.
```

</span>
```

Isso irá destruir o antigo `Counter` e remontar um novo.

> ### Elementos DOM de mesmo tipo {#dom-elements-of-the-same-type}

Reviewer 2: ```suggestion
### Elementos DOM de Mesmo Tipo {#dom-elements-of-the-same-type}
```

</span>
```

Isso irá destruir o antigo `Counter` e remontar um novo.

### Elementos DOM de mesmo tipo {#dom-elements-of-the-same-type}

Quando comparando dois Elementos DOM React do mesmo tipo, React olhará para os atributos de ambos, mantendo os nós DOM subjacentes e apenas atulizando os atributos modificados. Por exemplo:

```xml
<div className="before" title="stuff" />

<div className="after" title="stuff" />
```
Comparando estes dois elementos, o React sabe que deve modificar apenas o `className` no nó DOM subjacente.

Quando aqualizando o `style`, o React tambem sabe que deve atualizar apenas as propriedades modificadas. Por exemplo: 
```xml
<div style={{color: 'red', fontWeight: 'bold'}} />

<div style={{color: 'green', fontWeight: 'bold'}} />
```
Quando convertendo entre estes dois elementos, o React sabe que precisa modificar apenas o estilo `color`, mas não o `fontWeight`.

> Depois de manipular o nó DOM, o React itera recursivamente sobre os filhos.

Reviewer 3: ```suggestion
Depois de manipular o nó do DOM, o React itera recursivamente sobre os filhos.
```

> ### Elementos Componentes De Mesmo Tipo {#component-elements-of-the-same-type}

Reviewer 3: ```suggestion
### Componentes de Elementos do Mesmo Tipo {#component-elements-of-the-same-type}
```

> Quando um componente atualiza, a instance continua a mesma, então o estato é matido entre as renderizações. O React atualiza as props das instancias dos componente subjacentes para sincronizar com o novo elemento e então chama `componentWillReceiveProps()` e `componentWillUpdate()` na instancia subjacente.

Reviewer 3: ```suggestion
Quando um componente atualiza, a instância continua a mesma, então o estado é mantido entre as renderizações. O React atualiza as props das instâncias dos componente subjacentes para sincronizar com o novo elemento e então chama `componentWillReceiveProps()` e `componentWillUpdate()` na instância subjacente.
```

</ul>
```

O React irá sincronizar as duas árvores `<li>first</li>`, as duas árvores `<li>second</li>`, e então inserir a árvore `<li>third</li>`.

> Se você implementar ingenuamente, inserir um element no inicio trará uma performance menor. Por exemplo, a conversão entre essas duas árvores tem um baixo desempenho:

Reviewer 3: ```suggestion
Se você implementar ingenuamente, inserir um elemento no início trará uma performance menor. Por exemplo, a conversão entre essas duas árvores tem um desempenho baixo:
```

</ul>
```

Agora o React sabe que o elemento com a chave `'2014'` é o novo elemento, e os outros elementos com as chaves (_keys_) `'2015'` e `'2016'` apenas se moveram.

Na prática, achar uma chave (_key_) não é difícil. O elemento ao qual você irá exibir pode já possui um ID único, então a chave poderia ser gerada a partir do seu próprio dado:

```js
<li key={item.id}>{item.name}</li>
```

Quando não for o caso, você pode adicionar uma propriedade ID ao seu modelo ou utilizar um _hash_ em algumas partes do dado para gerar uma chave (_key_). A chave deve ser única apenas entre seus irmãos, e não única de forma global.

> Como um último recurso, você pode passar o index do item de um array como chave. Isso pode funcionar bem para itens que nunca são reordenados, mas reordená-los trará uma baixa performance.

Reviewer 2: ```suggestion
Como um último recurso, você pode passar o índice (_index_) do item de um array como chave. Isso pode funcionar bem para itens que nunca são reordenados, mas reordená-los trará uma baixa performance.
```

> Reordenar pode também causar um problema com o estado do componente quando os índices (indexes) são utilizados como chaves  (_keys_). A instância do componente é atualizada e reutilizada baseada na sua chave. Se a chave é um índice (index), mover o item modifica a chave. Como resultado disso, o estado do componente para coisas como _inputs_ não controlados podem ficar bagunçados e atualizar de uma forma inesperada.

Reviewer 2: ```suggestion
Reordenar pode também causar um problema com o estado do componente quando os índices (_indexes_) são utilizados como chaves  (_keys_). A instância do componente é atualizada e reutilizada baseada na sua chave. Se a chave é um índice (_index_), mover o item modifica a chave. Como resultado disso, o estado do componente para coisas como _inputs_ não controlados podem ficar bagunçados e atualizar de uma forma inesperada.
```

[Aqui](codepen://reconciliation/index-used-as-key) é um exemplo, no CodePen, de um problema que pode ser causado por usar índices como chaves, e [aqui](codepen://reconciliation/no-index-used-as-key) é uma versão atualizada do mesmo exemplo mostrando como a não utilização dos índices como chaves resolve os problemas relacionados a reordenação, ordenação e adição no início da lista.

> ## Compensações {#tradeoffs}

Reviewer 2: ```suggestion
## Compensações (_Tradeoffs_) {#tradeoffs}
```

É importante lembrar que o algoritmo de reconciliação é um detalhe de implementação. o React poderia re-renderizar o aplicativo inteiro a cada ação; o resultado final seria o mesmo. Apenas para ser claro, re-renderizar neste contexto significa chamar o método `render` para todos os componentes, isso não significa que o React irá desmontar e remontá-los. Isso significa apenas aplicar as diferenças seguindo as regras mencionadas nas seções anteriores.

Nós estamos regularmente refinando as heurísticas de modo a tornar mais rápidos os casos mais comuns. Na implementação atual, você pode dizer que uma subárvore foi movida para entre seus irmãos, mas não pode dizer que ela se moveu para qualquer outro lugar. O algoritmo irá re-renderizar toda a subárvore.

Devido ao React ser baseado em heurísticas, se as suposições por trás delas não encaixarem, a performance será menor.

1. O algoritmo não irá tentar sincronizar as subárvores de componentes de tipos diferentes. Se você perceber que está tentando alternar entre dois componentes de tipos diferentes com uma saída muito similar, você irá querer que eles tenham o mesmo tipo. Na prática nós não achamos que isso é um problema.

> 2. Chaves devem ser estáveis, previsíveis e únicas. Chaves estáveis (como as produzidas por `Math.random()`) irão causar a re-criação desnecessária de várias instâncias de componentes e nós DOM, o que pode causar uma degradação na performance e a perca do estado nos componentes filhos.

Reviewer: Corrigir a tradução, de **`Chaves estáveis`** para **`Chaves instáveis`**

</ul>
```

> Agora o React sabe que o elemento com a chave `'2014'` é o novo elemento, e os outros elementos com as chaves `'2015'` e `'2016'` apenas se moveram.

Reviewer 3: ```suggestion
Agora o React sabe que o elemento com a chave `'2014'` é o novo elemento, e os outros elementos com as chaves (_keys_) `'2015'` e `'2016'` apenas se moveram.
```

> Na pratica, achar uma chave não é difícil. O elemento ao qual você irá exibir pode já possui um ID unico, então a chave poderia ser gerada a partir do seu proprio dado:

Reviewer 3: ```suggestion
Na pratica, achar uma chave (_key_) não é difícil. O elemento ao qual você irá exibir pode já possui um ID único, então a chave poderia ser gerada a partir do seu próprio dado:
```

```js
<li key={item.id}>{item.name}</li>
```

> Quando não é o caso, você pode adicionar uma propriedade ID ao seu modelo ou utilizar um hash em algumas partes do dado para gerar uma chave. A chave deve ser única apenas entre seus irmãos, não globalmente.

Reviewer 3: ```suggestion
Quando não for o caso, você pode adicionar uma propriedade ID ao seu modelo ou utilizar um _hash_ em algumas partes do dado para gerar uma chave (_key_). A chave deve ser única apenas entre seus irmãos, e não única de forma global.
```

> Como um ú ltimo recurso, você pode passar o index do item de um array como chave. Isso pode funcionar bem para items que nunca são reordenados, mas reordena-los trará uma baixa performance.

Reviewer 3: s/ú ltimo/último

> Reordenar pode tambem causar um problema com o estado do componente quando os indexes são utilizados como chaves. A instance do componente é atualizada e reusada baseada na sua chave. Se a chave é um index, mover o item modifica a chave. Como resultado disso, o estado do componente para coisas como inputs não controlados podem ficar bagunçados e atualizar de uma forma inesperada.

Reviewer 3: ```suggestion
Reordenar pode tambem causar um problema com o estado do componente quando os índices (indexes) são utilizados como chaves (_keys_). A instância do componente é atualizada e reutilizada baseada na sua chave. Se a chave é um índice (index), mover o item modifica a chave. Como resultado disso, o estado do componente para coisas como _inputs_ não controlados podem ficar bagunçados e atualizar de uma forma inesperada.
```

> [Aqui](codepen://reconciliation/index-used-as-key) é um exemplo, no CodePen, de um problema que pode ser causado por usar indexes como chaves, e [aqui](codepen://reconciliation/no-index-used-as-key) é uma versão atualizada do mesmo exemplo mostrando como a não utilização dos indices como chaves resolve os problemas relacionados a reordenação, ordenação e adição no início da lista.

Reviewer 3: ```suggestion
[Aqui](codepen://reconciliation/index-used-as-key) é um exemplo, no CodePen, de um problema que pode ser causado por usar índices como chaves, e [aqui](codepen://reconciliation/no-index-used-as-key) é uma versão atualizada do mesmo exemplo mostrando como a não utilização dos indices como chaves resolve os problemas relacionados a reordenação, ordenação e adição no início da lista.
```

## Compensações {#tradeoffs}

> É importante lembrar que o algoritmo de reconciliação é um detalhe de implementação. o React poderia re-renderizar o aplicativo inteiro a cada ação; o resultado final seria o mesmo. Apenas para ser claro, re-renderizar neste contexto significa chamar o método `render` para todos os componentes, isso não significa que o React irá desmontar e remonta-los. Isso significa apenas aplicar as diferenças seguindo as regras mencionadas nas seção anteriores.

Reviewer 3: ```suggestion
É importante lembrar que o algoritmo de reconciliação é um detalhe de implementação. O React poderia re-renderizar o aplicativo inteiro a cada ação; o resultado final seria o mesmo. Apenas para ser claro, re-renderizar neste contexto significa chamar o método `render` para todos os componentes, isso não significa que o React irá desmontar e remontá-los. Isso significa apenas aplicar as diferenças seguindo as regras mencionadas nas seção anteriores.
```
